src/samplers/remote_sampler.js

```diff
@@ -0,0 +1,126 @@
+// @flow
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import request from 'request';
+import ProbabilisticSampler from './probabilistic_sampler.js';
+import RateLimitingSampler from './ratelimiting_sampler.js';
+import NullLogger from '../logger.js';
+
+const DEFAULT_INITIAL_SAMPLING_RATE = 0.001;
+const SAMPLING_REFRESH_INTERVAL = 60000;
+const DEFAULT_SAMPLING_HOST = '0.0.0.0';
+const DEFAULT_SAMPLING_PORT = 5778;
+const PROBABILISTIC_STRATEGY_TYPE = 0;
+const RATELIMITING_STRATEGY_TYPE = 1;
+
+export default class RemoteControlledSampler {
+
+    _sampler: Sampler;
+    _callerName: string;
+    _logger: any;
+    _host: string;
+    _port: number;
+    _timeoutHandle: any;
+    _intervalHandle: any;
+    _onSamplerUpdate: Function;
+    _refreshInterval: number;
+
+    constructor(callerName: string,
+                options: any) {
+
+        this._callerName = callerName;
+        this._sampler = options.sampler || new ProbabilisticSampler(DEFAULT_INITIAL_SAMPLING_RATE);
+        this._logger = options.logger || new NullLogger();
+        this._refreshInterval = options.refreshInterval || SAMPLING_REFRESH_INTERVAL;
+
+        let randomDelay: number = Math.random() * this._refreshInterval;
+        this._onSamplerUpdate = options.onSamplerUpdate;
+        this._host = options.host || DEFAULT_SAMPLING_HOST;
+        this._port = options.port || DEFAULT_SAMPLING_PORT;
+
+        // setTimeout and setInterval have delays, so for testing purposes I allow an alternative path
+        // that immediately executes the _refreshSamplingStrategy
```

Comment:
why can we not keep the production code free of "for testing only" code? The tests can always be written like this (just needs to be converted into callback format):
```
rs = RemoteSampler(delay=1ms)
for i in range(1, 1000) {
  if rs.sampler has changed {
    break;
  }
  sleep(1ms)
}
assert rs.sampler has changed
```
under normal conditions this test will complete within a few ms.

Right now we're messing up public API of the sampler with a flag that's only used for testing.